## Exploring Hacker News Posts

Exploring posts with either Ask HN or Show HN.

Users submit Ask HN posts to ask the Hacker News community a specific question. Below are a couple examples:

> Ask HN: How to improve my personal website?
> Ask HN: Am I the only one outraged by Twitter shutting down share counts?
> Ask HN: Aby recent changes to CSS that broke mobile?

Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting. Below are a couple of examples:

> Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform'
> Show HN: Something pointless I made
> Show HN: Shanhu.io, a programming playground powered by e8vm

#### Questions
We'll compare these two types of posts to determine the following:
- Do Ask HN or Show HN receive more comments on average?
- Do posts created at a certain time receive more comments on average?


I'll be using the folowing kaggle [dataset](https://www.kaggle.com/hacker-news/hacker-news-posts)

Note: The dataset was reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions.





#### Opening file and creating hn var

In [11]:
from csv import reader

opened_file = open ('hacker_news.csv')
readed_file = reader(opened_file)
hn = list(readed_file)

#### display first 5 rows of the hn dataset

In [12]:
print (hn[:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


#### Separate header from data

In [13]:
headers = hn[0]
hn=hn[1:]
print (headers)
print ('----')
print (hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
----
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


##### Extracting Ask HN and Show HN Posts
Extract Ask HN and show HN posts.
Print the number of posts.

In [14]:
ask_posts = []
show_posts = []
other_posts = []

for item in hn:
    title=item[1]
    if title.lower().startswith('ask hn'):
        ask_posts.append(item)
    elif title.lower().startswith('show hn'):
        show_posts.append(item)
    else:
        other_posts.append(item)
        
print (len(ask_posts))
print (len(show_posts))
print (len(other_posts))


1744
1162
17194


#### Calculate the Average Number of Comments for Ask HN and Show HN Posts

In [17]:
total_ask_comments = 0
for ask in ask_posts:
    total_ask_comments += int(ask[4])
avg_ask_comments = total_ask_comments / len(ask_posts)
print ('Average ask comments: {:.2f}'.format(avg_ask_comments))

total_show_comments=0
for show in show_posts:
    total_show_comments += int(show[4])
avg_show_comments = total_show_comments / len(show_posts)
print ('Average show comments: {:.2f}'.format(avg_show_comments))



Average ask comments: 14.04
Average show comments: 10.32


From the previous calculation we see that in average, **Ask HN** posts receive more comments than the **Show HN** posts 

| Post Type | Average comments |
|----|:-----:|
| Ask HN | 14.04 |
| Show HN | 10.32 |

#### Finding the amount of ask posts and comments by hour created

Finding if ask posts created at a certain time are more likely to attract comments

In [34]:
import datetime as dt

result_list = []

for ask in ask_posts:
    result_list.append([ask[6],int(ask[4])])

counts_by_hour = {}
comments_by_hour = {}

date_format = "%m/%d/%Y %H:%M"

for result in result_list:
    date = dt.datetime.strptime(result[0],date_format)
    hour = date.strftime("%H")
    if hour in counts_by_hour:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += result[1]
    else:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = result[1]



print (sorted(counts_by_hour.items()))
print (sorted(comments_by_hour.items()))

[('00', 55), ('01', 60), ('02', 58), ('03', 54), ('04', 47), ('05', 46), ('06', 44), ('07', 34), ('08', 48), ('09', 45), ('10', 59), ('11', 58), ('12', 73), ('13', 85), ('14', 107), ('15', 116), ('16', 108), ('17', 100), ('18', 109), ('19', 110), ('20', 80), ('21', 109), ('22', 71), ('23', 68)]
[('00', 447), ('01', 683), ('02', 1381), ('03', 421), ('04', 337), ('05', 464), ('06', 397), ('07', 267), ('08', 492), ('09', 251), ('10', 793), ('11', 641), ('12', 687), ('13', 1253), ('14', 1416), ('15', 4477), ('16', 1814), ('17', 1146), ('18', 1439), ('19', 1188), ('20', 1722), ('21', 1745), ('22', 479), ('23', 543)]


#### Calculating the average number of comments for Ask HN Posts by Hour

In [35]:
avg_by_hour = []

for hour in counts_by_hour:
    avg_by_hour.append([hour,comments_by_hour[hour]/counts_by_hour[hour]])

print(avg_by_hour)

[['09', 5.5777777777777775], ['13', 14.741176470588234], ['10', 13.440677966101696], ['14', 13.233644859813085], ['16', 16.796296296296298], ['23', 7.985294117647059], ['12', 9.41095890410959], ['17', 11.46], ['15', 38.5948275862069], ['21', 16.009174311926607], ['20', 21.525], ['02', 23.810344827586206], ['18', 13.20183486238532], ['03', 7.796296296296297], ['05', 10.08695652173913], ['19', 10.8], ['01', 11.383333333333333], ['22', 6.746478873239437], ['08', 10.25], ['04', 7.170212765957447], ['00', 8.127272727272727], ['06', 9.022727272727273], ['07', 7.852941176470588], ['11', 11.051724137931034]]


#### Sortint and printing values from a list of lists


In [47]:
swap_avg_by_hour = []

for item in avg_by_hour:
    swap_avg_by_hour.append([item[1],item[0]])

print(swap_avg_by_hour)
print('----------------')

sorted_swap = sorted(swap_avg_by_hour,reverse=True)

print ("Top 5 Hours for Ask Post Comments")

time_format = "%H"

for item in sorted_swap[:5]:
    date = dt.datetime.strptime(item[1],time_format)
    hour = date.strftime("%H:%M")
    print('{0}: {1:.2f} average comments per post'.format(hour,item[0]))


[[5.5777777777777775, '09'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [16.796296296296298, '16'], [7.985294117647059, '23'], [9.41095890410959, '12'], [11.46, '17'], [38.5948275862069, '15'], [16.009174311926607, '21'], [21.525, '20'], [23.810344827586206, '02'], [13.20183486238532, '18'], [7.796296296296297, '03'], [10.08695652173913, '05'], [10.8, '19'], [11.383333333333333, '01'], [6.746478873239437, '22'], [10.25, '08'], [7.170212765957447, '04'], [8.127272727272727, '00'], [9.022727272727273, '06'], [7.852941176470588, '07'], [11.051724137931034, '11']]
----------------
Top 5 Hours for Ask Post Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


#### conclusions
Based on the information, the best times (EST) to create Ask posts and get comments are:

|Hour|Avg. Comments per post|
|---|:------:|
|15:00|38.59|
|02:00|23.81|
|20:00|21.52|
|16:00|16.80|
|21:00|16.01|